Add forecast classes

peanutfun opened the pull request:
Changes proposed in this PR:
- Add forecast variants for Hazard and Impact.
- Adapt ImpactCalc to return ImpactForecast when needed.
- Add tests.
- Add tutorial for forecast data.

This PR resolves a major portion of #1118. Note that sub-tasks have been marked as completed when merged into this branch.

### PR Author Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] Correct target branch selected (if unsure, select `develop`)
- [x] Descriptive pull request title added
- [x] Source branch up-to-date with target branch
- [x] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [x] [Tests][testing] updated
- [x] [Tests][testing] passing
- [ ] No new [linter issues][linter]
- [ ] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/development/Guide_Review.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/development/Guide_continuous_integration_GitHub_actions.html
[linter]: https://climada-python.readthedocs.io/en/latest/development/Guide_continuous_integration_GitHub_actions.html#static-code-analysis

peanutfun committed:
Forecast templates

peanutfun committed:
Clean slate for Forecast

Evelyn-M committed:
1127 - Forecast base class init (#1167)

* define base class
* add init test
* Fix #1127

---------

Co-authored-by: Valentin Gebhart <valentin.gebhart@usys.ethz.ch>
Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

peanutfun committed:
Impact forecast class (#1168)

* Add ImpactForecast
* Add unit tests

luseverin committed:
Hazard forecast class (#1171)

* Create hazardForecast base class
* Update datetime to timedelta for lead_time
* Add tests

---------

Co-authored-by: luseverin <luca.severino@usys.ethz.ch>
Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

ValentinGebhart committed:
Change leadtime from datetime to timedelta (#1172)

* change leadtime to timedelta
* adapt impactforecast docstrings and tests
* Remove stray comment

---------

Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

luseverin committed:
Hazard forecast test select (#1175)

* Add HazardForecast.select test

---------

Co-authored-by: luseverin <luca.severino@usys.ethz.ch>
Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

luseverin committed:
Hazard forecast concat test (#1178)

* Add test for hazardForecast.concat
* Skip HazardForecast.concat test

---------

Co-authored-by: luseverin <luca.severino@usys.ethz.ch>
Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

peanutfun committed:
Check if `Impact.select` works on `ImpactForecast` (#1170)

* Add test for ImpactForecast.select
* minor test modification

---------

Co-authored-by: Evelyn Mühlhofer <evelyn.muehlhofer@meteoswiss.ch>

peanutfun committed:
Check if `Impact.concat` works on `ImpactForecast` (#1174)

* Add test stub for ImpactForecast.concat
* Use correct command for skipping a test
* Fix merge issue

ValentinGebhart committed:
adapt from_hdf5 and write_hdf5

peanutfun committed:
Check sizes of forecast data vs. Hazard and Impact (#1180)

* Add data shape check for HazardForecast, ImpactForecast

ValentinGebhart committed:
implement code review

ValentinGebhart committed:
add comment to test

ValentinGebhart committed:
Merge pull request #1181 from CLIMADA-project/adapt_hdf5_write_from

Adapt Hazard from_hdf5 and write_hdf5 to forecast attributes

chahank committed:
Forecast/select extended tests (#1182)

* Add extended select impact forecast test
* Add extended hazard forecast test

---------

Co-authored-by: Chahan Kropf <chahan.kropf@posteo.com>
Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

luseverin committed:
Impact calc return impact forecast (#1179)

* Return impactForecast object in _return_impact
* Return impactForecast object in _return_empty
* Add full  impactcalc test for impactForecast
* Raise value error when computing impact with impact forecast without saving impact matrix
* Cosmetics: Improve error message, move test to own class
* Write nans for eai_exp and aai_agg when forecast is used

---------

Co-authored-by: luseverin <luca.severino@usys.ethz.ch>
Co-authored-by: Chahan Kropf <chahan.kropf@posteo.com>
Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

elianekobler committed:
add idx boolean selection for member and leadtime (#1183)

* add idx boolean selection for member and leadtime
* Add test for weird data types

---------

Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

ValentinGebhart committed:
Add select to ImpactForecast (#1188)

* Add ImpactForecast.select

---------

Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

ValentinGebhart committed:
Add select method to HazardForecast (#1185)

* add HazardForecast.select and extent test

---------

Co-authored-by: Eliane Kobler <ekobler@student.ethz.ch>
Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

peanutfun committed:
Make Impact.concat support ImpactForecast

peanutfun committed:
Add `HazardForecast.concat` (#1184)

* Add HazardForecast.concat

luseverin committed:
Implement mean min max for impact forecast and hazard forecast (#1187)

Add min/mean/max methods

---------

Co-authored-by: luseverin <luca.severino@usys.ethz.ch>
Co-authored-by: Chahan Kropf <chahan.kropf@posteo.com>
Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

peanutfun committed:
Support `ImpactForecast.concat` (#1192)

* Make Impact.concat support ImpactForecast
* Add tests

ValentinGebhart committed:
Add hdf5 IO for ImpactForecast (#1190)

* implement changes in Impact IO and ImpactForecast IO
* Update tests
* Fix small issue when reading H5 for ImpactForecast

---------

Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

luseverin committed:
Implement quantiles and median for hazard and impact forecasts (#1191)

* Add quantile and median for hazard and impact forecast
* Review

---------

Co-authored-by: luseverin <luca.severino@usys.ethz.ch>
Co-authored-by: Chahan Kropf <chahan.kropf@posteo.com>
Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>

ValentinGebhart committed:
Merge branch 'forecast-class' of github.com:CLIMADA-project/climada_python into forecast-class

ValentinGebhart committed:
Add forecast tutorial (#1193)

* Add impact forecast tutorial
* Add 'dim' argument to reductions
* Update tests

---------

Co-authored-by: Lukas Riedel <34276446+peanutfun@users.noreply.github.com>
Co-authored-by: luseverin <luca.severino@usys.ethz.ch>
Co-authored-by: Evelyn Mühlhofer <evelyn.muehlhofer@meteoswiss.ch>
Co-authored-by: Chahan Kropf <chahan.kropf@posteo.com>

ValentinGebhart committed:
Merge branch 'forecast-class' of github.com:CLIMADA-project/climada_python into forecast-class

peanutfun commented on climada/engine/test/test_impact_calc.py:
```diff
@@ -603,7 +692,47 @@ def test_single_exp_zero_mdr(self):
+    @pytest.fixture
+    def impact_calc_forecast(self, impact_calc):
+        """Write NaNs to attributes that are not used"""
+        impact_calc["aai_agg"] = np.full_like(impact_calc["aai_agg"], np.nan)
```
`aai_agg` is a scalar
```suggestion
        impact_calc["aai_agg"] = np.nan
```

peanutfun committed:
Apply suggestion from @peanutfun

peanutfun committed:
Update CHANGELOG.md

peanutfun commented:
@Evelyn-M How was the `fcst_icon_seamless_2024-06-18_12_00.nc` test file created? I see that it's in a different CRS than the "latest" example files downloaded in the tutorial.

Evelyn-M commented:
> @Evelyn-M How was the `fcst_icon_seamless_2024-06-18_12_00.nc` test file created? I see that it's in a different CRS than the "latest" example files downloaded in the tutorial.

I downloaded it from the meteoswiss-internal data storage, since the API only permits downloads until -24 hrs.

peanutfun commented:
So it had been reprojected? Does MCH store the data differently than OGD? I am just wondering because it might be important to highlight the caveats when importing datasets with varying CRS and coordinate specifications.

Evelyn-M commented:
I suppose so, yes. One is the MCH grid commonly used internally for national gridded forecast products, and the OGD data has a different grid and a much larger rectangular extent around Switzerland.

ValentinGebhart commented on climada/util/forecast.py:
```diff
@@ -0,0 +1,193 @@
+import numpy as np
+
+
+class Forecast:
```
It seems that there are people recommending to add "Mixin" to the class name for mixin classes. E.g. "ForecastMixin"

peanutfun replied:
Done

ValentinGebhart commented on climada/util/forecast.py:
```diff
@@ -0,0 +1,193 @@
+    """
+    concat_kws = {} if concat_kws is None else concat_kws
+    return forecast.concat(
+        [
```
I found this a bit hard to read, to understand what is happening. I see that the code is doing what it should, but maybe the variable names could be changed to make things clearer?
E.g.:
values --> attr_values
select --> attr_name
reduce_attr --> reduce_method

peanutfun replied:
Renamed and reworked a bit

ValentinGebhart commented on climada/util/test/test_forecast.py:
```diff
@@ -0,0 +1,95 @@
+    npt.assert_array_equal(idx, np.array([True, False, False, False]), strict=True)
+
+    idx = forecast.idx_lead_time(None)
+    npt.assert_array_equal(idx, np.array([False, False, False, False]), strict=True)
```
no tests for 
_reduce_attrs
reduce_unique_selection
_unique_or_default

but the working of these might be tested in the functions calling them

peanutfun replied:
Private methods cannot be tested separately (because they are private?). `reduce_unique_selection` is tested via the reduction methods of the final classes. It's correct that a unit test should check the standalone function

peanutfun replied:
Since `reduce_unique_selection` is only used within the reduction methods, we decided against a dedicated unit test for the function

ValentinGebhart commented on climada/util/forecast.py:
```diff
@@ -0,0 +1,193 @@
+        return reduced_attrs
+
+
+def reduce_unique_selection(
```
on first reading I found the back on forth confusing. e.g. HazardForecast.min() calls reduce_unqiue_slection with again calls HazardForecast.min(). 

I see why this works, just to comment on readability

ValentinGebhart commented on climada/hazard/forecast.py:
```diff
@@ -0,0 +1,511 @@
+        Returns
+        -------
+        HazardForecast
+            A HazardForecast object with the min intensity and fraction.
```
typo: min should be max

ValentinGebhart commented on climada/hazard/forecast.py:
```diff
@@ -0,0 +1,511 @@
+
+    def min(self, dim: Literal["member", "lead_time"] | None = None):
+        """
+        Reduce the intensity and fraction of a HazardForecast to the minimum
```
in these functions (min, max, qunatile, mean, also for ImapctForecast), should we include that we reduce for each centroid? Might be confusing if the users thinks we reduce e.g. the whole intensity matrix

peanutfun replied:
I added a warning that intensity and fraction are reduced independently for each centroid.

ValentinGebhart commented on climada/hazard/forecast.py:
```diff
@@ -0,0 +1,511 @@
+        Returns
+        -------
+        HazardForecast
+            A HazardForecast object with the min intensity and fraction.
```
typo: min should be mean

peanutfun replied:
Done

ValentinGebhart commented on climada/hazard/forecast.py:
```diff
@@ -0,0 +1,511 @@
+
+    def quantile(self, q: float, dim: Literal["member", "lead_time"] | None = None):
+        """
+        Reduce the impact matrix and at_event of a HazardForecast to the quantile value.
```
here it should be "intensity matrix", not impact? what is at_event?

peanutfun replied:
Copy-paste error from ImpactForecast

ValentinGebhart commented on climada/hazard/forecast.py:
```diff
@@ -0,0 +1,511 @@
+        HazardForecast
+            A HazardForecast object with the quantile intensity and fraction.
+        """
+        return self._quantile(q=q, dim=dim)
```
why do we have a private function for this? we just directly call it?

ValentinGebhart replied:
ok maybe it is because of the "event_name" that you want to input? then ok

peanutfun replied:
Correct, passing different `event_name` parameters for `median` and `quantile` is the (only) reason.

ValentinGebhart commented on climada/hazard/forecast.py:
```diff
@@ -0,0 +1,511 @@
+        event_name: str | None = None,
+    ):
+        """
+        Reduce the impact matrix and at_event of a HazardForecast to the quantile value.
```
impact matrix and at_event?

peanutfun replied:
Fixed

ValentinGebhart commented on climada/hazard/forecast.py:
```diff
@@ -0,0 +1,511 @@
+
+    def median(self, dim: Literal["member", "lead_time"] | None = None):
+        """
+        Reduce the impact matrix and at_event of a HazardForecast to the median value.
```
impact matrix and at_event?

peanutfun replied:
Fixed

ValentinGebhart commented on climada/hazard/forecast.py:
```diff
@@ -0,0 +1,511 @@
+            **self._reduce_attrs("mean"),
+        )
+
+    # TODO: Do not densify the entire matrix but compute quantiles column-wise!
```
not yet implemented

peanutfun replied:
See https://github.com/CLIMADA-project/climada_python/issues/1203

ValentinGebhart commented on climada/hazard/forecast.py:
```diff
@@ -0,0 +1,511 @@
+            f"lt_{lt / np.timedelta64(1, 'h'):.0f}h_m_{m}"
+            for lt, m in zip(kwargs["lead_time"], kwargs["member"])
+        ]
+        kwargs["date"] = np.zeros_like(kwargs["date"], dtype=int)
```
Is it on purpose that we set date to zero? Could be the date of the forecast init?

peanutfun replied:
That would interfere with the definition of `date` in `Hazard`, where it describes the date of the event itself. The forecast init date might also be difficult to determine, because different data sources might define the quantity differently.

ValentinGebhart commented on climada/hazard/xarray.py:
```diff
@@ -238,6 +238,8 @@ class HazardXarrayReader:
     ----------
     data : xr.Dataset
         The data to be read as hazard.
+    data_stacked : xr.Dataset
```
not entirely sure why you introduce data_stacked here? It seems that you use it in exactly the same way as you used data before?

peanutfun replied:
Because we want to load more data from the stacked data object here: https://github.com/CLIMADA-project/climada_python/blob/380f0dd71ad59e6644bd625599eb46c01ce31463/climada/hazard/forecast.py#L492

I agree it's not elegant, though.

ValentinGebhart commented on climada/engine/test/test_impact_forecast.py:
```diff
@@ -0,0 +1,471 @@
+@pytest.mark.parametrize("dense", [True, False])
+def test_write_read_hdf5(impact_forecast, tmp_path, dense):
+
+    file_name = tmp_path / "test_hazard_forecast.h5"
```
```suggestion
    file_name = tmp_path / "test_impact_forecast.h5"
```

ValentinGebhart commented on climada/engine/impact_calc.py:
```diff
@@ -230,12 +232,31 @@ def _return_impact(self, imp_mat_gen, save_mat):
+            if isinstance(self.hazard, HazardForecast):
+                eai_exp = np.full_like(eai_exp, np.nan, dtype=eai_exp.dtype)
+                aai_agg = np.nan
+                LOGGER.warning(
```
Do we want to warn this every time? Maybe makes sense?

peanutfun replied:
Might indeed be superfluous. Do you have a strong opinion here, @chahank ?

chahank replied:
No strong opinion, slight tendency do superfluous.

ValentinGebhart commented on climada/engine/test/test_impact_calc.py:
```diff
@@ -46,17 +50,102 @@
+@pytest.fixture
+def impact_calc(exposure, hazard):
+    imp_mat = np.ones((len(hazard.event_id), exposure.gdf.shape[0]))
+    aai_agg = np.sum(exposure.gdf["value"]) * hazard.frequency[0]
```
I understand that all hazard events affect all expsoure centroids maximally? Wouldn't that mean that 
aai_agg = np.sum(exposure.gdf["value"]) * np.sum(hazard.frequency)?

chahank replied:
I think this is correct. Why not have the acutal impact calculation here? This seems quite flimsy if anything changes in exposure or hazard.

peanutfun replied:
@chahank This is a fixture for comparing the actual impact calculation to a pre-defined result. I think it would not make sense to compare one result of an impact calc against another, right?

@ValentinGebhart You are right, but then the namesake average is taken, which is the same as multiplying with `hazard.frequency[0]` because all frequencies and intensities are the same. I think it should be

```suggestion
    aai_agg = np.sum(np.mean(exposure.gdf["value"] * hazard.frequency[:, np.newaxis], axis=0))
```

peanutfun replied:
Sorry, it's
```suggestion
    aai_agg = np.sum(exposure.gdf["value"]) * np.mean(hazard.frequency)
```

ValentinGebhart replied:
sorry, still confused here about why you would compute the mean on not the sum over the frequencies..

Isn't the aai_agg = \sum_{event, loc} impact_{evect, loc} * frequency_{event}?
The sum over loc would give np.sum(exposure.gdf["value"]) and the sum over events would give np.sum(hazard.frequency). 

But maybe I miss something here, we can chat in person

peanutfun replied:
Indeed, fixed now.

ValentinGebhart commented on climada/engine/test/test_impact_calc.py:
```diff
@@ -46,17 +50,102 @@
+def impact_calc(exposure, hazard):
+    imp_mat = np.ones((len(hazard.event_id), exposure.gdf.shape[0]))
+    aai_agg = np.sum(exposure.gdf["value"]) * hazard.frequency[0]
+    eai_exp = np.ones(exposure.gdf.shape[0]) * hazard.frequency[0]
```
similar, shouldn't this be
eai_exp = np.ones(exposure.gdf.shape[0]) * np.sum(hazard.frequency)?

ValentinGebhart replied:
same here, for me still the sum over frequencies

peanutfun replied:
Fixed now

chahank commented:
Can you please let me know once this is final for review?

ValentinGebhart commented on climada/hazard/test/test_forecast.py:
```diff
@@ -0,0 +1,589 @@
+    ):
+        """Check if Hazard.select works on the derived class"""
+
+        select_mask = np.array([3, 2])
```
why do we separate between `select_mask` and `ordered_select_mask` if they have the same values?

peanutfun replied:
This is a relict from porting the test over from ImpactForecast. In `Impact`,  the select algorithm preserves the original event order, irrespective of the order passed to the `select` method

ValentinGebhart commented on climada/hazard/test/test_forecast.py:
```diff
@@ -0,0 +1,589 @@
+        select_mask = np.array([3, 2])
+        ordered_select_mask = np.array([3, 2])
+        if var == "date":
+            # Date needs to be a valid delta
```
can you clarify comment? Like this it is not clear to mewhat `Date` needs

peanutfun replied:
Done. I was confused as well. `select(date=X)` only supports that X is a sequence of two values, with the first indicating the begin date and the second indicating the end date. If they are in inverse order, no date matches. I updated the test accordingly to make this more clear.

ValentinGebhart commented on climada/hazard/test/test_forecast.py:
```diff
@@ -0,0 +1,589 @@
+
+        # Check centroids
+        assert len(haz_fc.centroids.lat) == expected_n_centroids
+        assert len(haz_fc.centroids.lon) == expected_n_centroids
```
It seems we are only checking if everything has the right shape. Should we also check (by a few samples) that the intensity value for a given (member, lead_time) is assigned to the correct member and lead_time in the hazard forecast?

peanutfun replied:
Added a test for the actual coordinates, but technically this is not changed in the derived class and is tested in the base class tests.

ValentinGebhart commented on climada/hazard/test/test_forecast.py:
```diff
@@ -0,0 +1,589 @@
+        expected_n_events = (
+            forecast_netcdf_file["n_eps"] * forecast_netcdf_file["n_lead_time"]
+        )
+        npt.assert_array_equal(haz_fc.date, np.zeros(expected_n_events, dtype=int))
```
what about setting the date to the date of the reference time? Maybe it makes sense not to, just thinking

peanutfun replied:
See comment above

ValentinGebhart commented on climada/engine/impact_forecast.py:
```diff
@@ -0,0 +1,493 @@
+
+    @at_event.setter
+    def at_event(self, value):
+        """Set the total impact for each member/lead_time combination."""
```
was this getter/setter set to allow for some future type checking or so? otherwise maybe not necessary?

peanutfun replied:
Getter/setter methods exist to emit a warning when used

ValentinGebhart commented on climada/engine/test/test_impact_forecast.py:
```diff
@@ -0,0 +1,471 @@
+        select_mask = np.array([2, 1])
+        ordered_select_mask = np.array([1, 2])
+        if var == "date":
+            # Date needs to be a valid delta
```
same as in test_hazard_forecast.py

peanutfun replied:
Fixed

ValentinGebhart commented on climada/engine/test/test_impact_forecast.py:
```diff
@@ -0,0 +1,471 @@
+        )
+
+        num_centroids = len(impact_kwargs["coord_exp"])
+        imp_fc_select = impact_forecast.select(event_names=["aaaaa", "foo"])
```
Is it on purpose that in hazard_forecast.select() we raise an error when there is an emtpy selection and here not?

peanutfun replied:
This follows the respective base class implementations 🤷

ValentinGebhart commented on climada/engine/test/test_impact_forecast.py:
```diff
@@ -0,0 +1,471 @@
+    @pytest.fixture
+    def reduction_results_dim(self):
+        return {
+            "lead_time": {
```
I think `at_event` should always return the sum of the impact matrix over the locations (i.e., axis=1), independent of if the eventdimension corresponds to lead_time or member. 

In the test, this is tested for "member" but not for "lead_time". The values look like they would give the same sum(axis=1) and sum(axis=0). Maybe we could adapt a value such that we know the right axis is summed? If too compilated, maybe the test for "member" is enough.

peanutfun replied:
I don't understand. `at_event` is always exactly `np.sum(imp_mat, axis=1)`, both for `member` and for `lead_time`.

> I think at_event should always return the sum of the impact matrix over the locations (i.e., axis=1), ...

I agree! The test checks this explicitly, in my perception?

ValentinGebhart replied:
sorry, my comment was not clear. You are right, `at_event` should always be exactly `np.sum(imp_mat, axis=1)`. 

My point was that in the test of "lead_time", the test values seem to be chosen such that the impact matrix is always symmetric (off diagonals are the same), which is why `np.sum(imp_mat, axis=1)` is the same as `np.sum(imp_mat, axis=0)`. So we actually do not test if the correct axis is summed here. However, we test this in the test of "member", so that should suffice. For me we can leave as is.

peanutfun replied:
@ValentinGebhart I updated the test. The resulting matrices are now not symmetric anymore, making the test more clear.

ValentinGebhart reviewed (commented):
Looks very good to me, all comments are minor I think. 

I have not yet reviewed the tutorial, will do a separate review for that when i have time.

peanutfun commented:
@chahank This is final for review 🙂

chahank commented:
Excellent! I will try to do it swiftly.

chahank commented on climada/engine/test/test_impact.py:
```diff
@@ -47,26 +47,30 @@
+        "aai_agg": 14.4,
+        "unit": "USD",
+        "frequency_unit": "1/month",
+        "imp_mat": sparse.csr_matrix(
```
Is it worth to test for negative impact values too?

ValentinGebhart commented on doc/user-guide/climada_engine_ImpactForecast.ipynb:
After downloading the data, we do 
````
dict_ens_data["TOT_PREC"] = time_ops.delta(
    dict_ens_data["TOT_PREC"], np.timedelta64(1, "h")
)
````
This looks like we change from total values to differences. I assume we do this because how TOT_PREC is defined?

We should for sure add an explanation and warning because if I just copy the code and use it for wind speed (VMAX_10M) I do NOT want to compute the difference.

peanutfun replied:
Indeed, I added a comment

chahank commented on climada/hazard/test/test_base.py:
```diff
@@ -46,27 +46,29 @@
+        "haz_type": "TC",
+        "pool": None,
+        "intensity": sparse.csr_matrix(
+            [[0.2, 0.3, 0.4], [0.1, 0.1, 0.01], [4.3, 2.1, 1.0], [5.3, 0.2, 0.0]]
```
Please include negative value intensity.

peanutfun replied:
The change to this file only wraps the previous `dummy_hazard` function into a fixture. I think changing the intensity values is out of scope, because changes here might have effects on all other hazard tests. We can pick this up in a discussion about the `conftest.py` fixtures, see #1278

chahank commented:
Is there a plan already to work on the comments by @ValentinGebhart?

chahank commented:
The tutorial is now titled only impact forecast, but it is mostly about hazard forecast. This is quite confusing. Can this maybe be split into two tutorials? Or give it a different name?

chahank commented:
A few comments on the tutorial:

- Can the tutorial of the impact forecast please add a section a the top that shows the whole pipeline within one cell for a fast introduction? 

- I think the cells with `?ImpactForecast` and `?HazardForecast` do not serve a clear purpose.

- OVerall, the tutorial could need a bit of improvements in clarity. Some of the cells are just there without explanations.

- There are a lot of of Logging that indicate something should be improved in the code I think. It is unclear why `imp_fct_all_mem_1lt = impact_forecast.max(dim="lead_time")` triggers many lines of info and warning logging. Or in some cell there is several repetitions of `2026-01-08 14:55:43,321 - climada.engine.impact - INFO - The eai_exp and aai_agg are computed for the selected subset of events WITHOUT modification of the frequencies.`.

chahank reviewed (changes requested):
Excellent addition to the CLIMADA platform! 

Please find a few detailled comments by @ValentinGebhart to address, and some user friendliness requests for the tutorial.

peanutfun committed:
Merge branch 'develop' into forecast-class

peanutfun committed:
Rename 'Forecast' to 'ForecastMixin'

peanutfun committed:
Rework parameters to 'reduce_unique_selection'

peanutfun committed:
Fix inconsistent override of Impact.from_hdf5

peanutfun committed:
Update docstrings

peanutfun committed:
Fix overrides in HazardForecast

peanutfun committed:
Fix more docstrings

peanutfun committed:
Update climada/engine/test/test_impact_forecast.py

Co-authored-by: Valentin Gebhart <60438839+ValentinGebhart@users.noreply.github.com>

peanutfun committed:
Avoid warnings emitted from ImpactForecast.select

peanutfun committed:
Update tests

peanutfun committed:
Merge branch 'forecast-class' of github.com:CLIMADA-project/climada_python into forecast-class

peanutfun committed:
Try suppressing more warnings

peanutfun committed:
Update tutorial

peanutfun committed:
Update tutorial

peanutfun committed:
Fix impact_calc fixture

peanutfun committed:
Fix symmetry issue in ImpactForecast tests

peanutfun commented:
@chahank Thanks for the review!

> Can the tutorial of the impact forecast please add a section a the top that shows the whole pipeline within one cell for a fast introduction?

I added that.

> I think the cells with ?ImpactForecast and ?HazardForecast do not serve a clear purpose.

Removed the cells.

> OVerall, the tutorial could need a bit of improvements in clarity. Some of the cells are just there without explanations.

Can you be more specific? What is missing apart from some "flavor text"?

> There are a lot of of Logging that indicate something should be improved in the code I think. It is unclear why imp_fct_all_mem_1lt = impact_forecast.max(dim="lead_time") triggers many lines of info and warning logging. Or in some cell there is several repetitions of 2026-01-08 14:55:43,321 - climada.engine.impact - INFO - The eai_exp and aai_agg are computed for the selected subset of events WITHOUT modification of the frequencies..

I suppressed more warnings where it seemed appropriate. The particular warning you mention is now suppressed when computing an aggregate, but not when calling `.select` directly (where the warning originates from)

chahank commented:
> @chahank Thanks for the review!
> 
> > Can the tutorial of the impact forecast please add a section a the top that shows the whole pipeline within one cell for a fast introduction?
> 
> I added that.

Great, thanks!

> 
> > I think the cells with ?ImpactForecast and ?HazardForecast do not serve a clear purpose.
> 
> Removed the cells.

Thanks!


> 
> > OVerall, the tutorial could need a bit of improvements in clarity. Some of the cells are just there without explanations.
> 
> Can you be more specific? What is missing apart from some "flavor text"?
> 
> > There are a lot of of Logging that indicate something should be improved in the code I think. It is unclear why imp_fct_all_mem_1lt = impact_forecast.max(dim="lead_time") triggers many lines of info and warning logging. Or in some cell there is several repetitions of 2026-01-08 14:55:43,321 - climada.engine.impact - INFO - The eai_exp and aai_agg are computed for the selected subset of events WITHOUT modification of the frequencies..
> 
> I suppressed more warnings where it seemed appropriate. The particular warning you mention is now suppressed when computing an aggregate, but not when calling `.select` directly (where the warning originates from)

Excellent.

There still is the warning `2026-04-10 14:03:38,318 - climada.engine.impact - WARNING - The Impact.tot_value attribute is deprecated. Use Exposures.affected_total_value to calculate the affected total exposure value based on a specific hazard intensity threshold` which is triggered`

This is due to the copying of `tot_value`. I think this can be quite confusing to the user.

peanutfun committed:
Merge branch 'develop' into forecast-class

peanutfun committed:
Avoid Impact.tot_value warnings. Fix data read.

peanutfun commented:
> This is due to the copying of tot_value. I think this can be quite confusing to the user.

Fixed now.

For some reason, the latest version of xarray does not seem to read the `lead_time` coordinate of the demonstration dataset on the data API correctly. I wrote a quick function to fix it and added it to the tutorial. Loading the latest data from the OGD API still works without issues. Not sure what exactly changed, but there seems to be an ongoing change about how lead times are decoded, with lots of regressions on the way. I think we can leave it as is.

I think this is ready to merge now.
